modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointDirtyPages.java

```diff
@@ -108,33 +100,54 @@ public IgniteConcurrentMultiPairQueue<PersistentPageMemory, GroupPartitionId> to
      * @param partId Partition ID.
      */
     public @Nullable CheckpointDirtyPagesView findView(int grpId, int partId) {
-        if (dirtyPages.isEmpty()) {
-            return null;
+        for (int i = 0; i < dirtyPages.size(); i++) {
+            CheckpointDirtyPagesView view = findView(i, grpId, partId);
+
+            if (view != null) {
+                return view;
+            }
         }
 
-        FullPageId startPageId = new FullPageId(pageId(partId, (byte) 0, 0), grpId);
-        FullPageId endPageId = new FullPageId(pageId(partId + 1, (byte) 0, 0), grpId);
+        return null;
+    }
 
+    /**
+     * Looks for dirty page IDs views for a specific group and partition.
+     *
+     * @param pageMemory Page memory.
+     * @param grpId Group ID.
+     * @param partId Partition ID.
+     */
+    public @Nullable CheckpointDirtyPagesView findView(PersistentPageMemory pageMemory, int grpId, int partId) {
         for (int i = 0; i < dirtyPages.size(); i++) {
-            FullPageId[] pageIds = dirtyPages.get(i).dirtyPages.pageIds;
+            if (dirtyPages.get(i).pageMemory == pageMemory) {
+                return findView(i, grpId, partId);
+            }
+        }
 
-            int fromIndex = binarySearch(pageIds, startPageId, DIRTY_PAGE_COMPARATOR);
+        return null;
+    }
 
-            fromIndex = fromIndex >= 0 ? fromIndex : Math.min(pageIds.length - 1, -fromIndex - 1);
+    private @Nullable CheckpointDirtyPagesView findView(int dirtyPagesIdx, int grpId, int partId) {
+        FullPageId startPageId = new FullPageId(pageId(partId, (byte) 0, 0), grpId);
+        FullPageId endPageId = new FullPageId(pageId(partId + 1, (byte) 0, 0), grpId);
 
-            if (!equalsByGroupAndPartition(startPageId, pageIds[fromIndex])) {
-                continue;
-            }
+        FullPageId[] pageIds = dirtyPages.get(dirtyPagesIdx).dirtyPages.pageIds;
 
-            int toIndex = binarySearch(pageIds, fromIndex, pageIds.length, endPageId, DIRTY_PAGE_COMPARATOR);
+        int fromIndex = binarySearch(pageIds, startPageId, DIRTY_PAGE_COMPARATOR);
 
-            // toIndex cannot be 0 because endPageId is greater than startPageId by DIRTY_PAGE_COMPARATOR.
-            toIndex = toIndex > 0 ? toIndex - 1 : -toIndex - 2;
+        fromIndex = fromIndex >= 0 ? fromIndex : Math.min(pageIds.length - 1, -fromIndex - 1);
 
-            return new CheckpointDirtyPagesView(i, fromIndex, toIndex);
+        if (!equalsByGroupAndPartition(startPageId, pageIds[fromIndex])) {
+            return null;
         }
 
-        return null;
+        int toIndex = binarySearch(pageIds, fromIndex, pageIds.length, endPageId, DIRTY_PAGE_COMPARATOR);
+
+        // toIndex cannot be 0 because endPageId is greater than startPageId by DIRTY_PAGE_COMPARATOR.
+        toIndex = toIndex > 0 ? toIndex : -toIndex - 1;
```

Comment:
This comment wouldn't be required if you wrote ">= 0". Why does it have to be more complicated?

Reply:
Fix it